File: components/Select.js
(function () {
  const closeIcon = "\u2715";

  const selectContainer = document.querySelector(".custom-select");
  const isMultiple = selectContainer.getAttribute("data-multiple") === "true";
  const arrow = selectContainer.querySelector(".arrow");
  const placeholderTextContainer =
    selectContainer.querySelector(".placeholder");
  const selectedValueContainer = selectContainer.querySelector(
    ".select-value-container"
  );

  const placeholderText = selectContainer.getAttribute("data-placeholder");
  const selectName = selectContainer.getAttribute("data-name");

  let multiSelectValues = [];

  function replaceIcon() {
    if (!selectedValueContainer.classList.contains("hidden")) {
      arrow.classList.remove("arrow");
      arrow.classList.add("remove-icon");
      arrow.innerHTML = closeIcon;
    } else {
      arrow.classList.remove("remove-icon");
      arrow.classList.add("arrow");
      arrow.innerHTML = "";
    }
  }

  selectContainer.addEventListener("click", (e) => {
    e.preventDefault();
    e.stopPropagation();
    // Toggle dropdown
    if (e.target.matches(".select-button")) {
      selectContainer.classList.toggle("active");
    }

    if (e.target.matches(".select-dropdown .option")) {
      if (isMultiple) {
        const label = e.target.childNodes[0].textContent;
        const value = e.target.getAttribute("data-value");
        const index = multiSelectValues.indexOf(value);

        // Create Badges
        const createSelectedValue = document.createElement("span");
        const closeIconElement = document.createElement("span");

        closeIconElement.classList.add("remove-icon");
        createSelectedValue.classList.add("selected-value");
        createSelectedValue.setAttribute("data-value", value);
        createSelectedValue.textContent = label;
        closeIconElement.textContent = closeIcon;
        createSelectedValue.appendChild(closeIconElement);

        // Create Checkbox
        const checkbox = document.createElement("input");
        checkbox.type = "checkbox";
        checkbox.name = selectName;
        checkbox.value = value;
        checkbox.checked = true;

        if (index === -1) {
          multiSelectValues.push(value);
          e.target.classList.add("checked");
          createSelectedValue.appendChild(checkbox);
          selectedValueContainer.appendChild(createSelectedValue);
        } else {
          multiSelectValues.splice(index, 1);
          e.target.classList.remove("checked");
          const selectedValue = selectedValueContainer.querySelector(
            `.selected-value[data-value='${value}']`
          );
          selectedValueContainer.removeChild(selectedValue);
        }
        selectedValueContainer.classList.remove("hidden");
        placeholderTextContainer.classList.add("hidden");
      } else {
        if (!selectedValueContainer.classList.contains("hidden")) {
          return;
        }
        const label = e.target.childNodes[0].textContent;
        const value = e.target.getAttribute("data-value");
        // create a radio button
        const radio = document.createElement("input");
        radio.type = "radio";
        radio.name = "select";
        radio.value = value;
        radio.checked = true;

        // create a label
        const labelElement = document.createElement("label");
        labelElement.textContent = label;

        labelElement.appendChild(radio);
        selectedValueContainer.textContent = "";
        selectedValueContainer.appendChild(labelElement);
        e.target.classList.add("checked");
        placeholderTextContainer.classList.add("hidden");
        selectedValueContainer.classList.remove("hidden");
        replaceIcon();
      }
    }

    // Removing selected value
    if (e.target.matches(".remove-icon")) {
      if (isMultiple) {
        const selectedValue = e.target.parentNode;
        const value = selectedValue.getAttribute("data-value");
        const index = multiSelectValues.indexOf(value);
        const option = selectContainer.querySelector(
          `.option[data-value='${value}']`
        );
        option.classList.remove("checked");
        multiSelectValues.splice(index, 1);
        selectedValueContainer.removeChild(selectedValue);
        if (multiSelectValues.length < 1) {
          selectedValueContainer.classList.add("hidden");
          placeholderTextContainer.classList.remove("hidden");
        }
      } else {
        const selectedOption = selectedValueContainer.querySelector(
          "input[type='radio']"
        );
        const option = selectContainer.querySelector(
          `.option[data-value='${selectedOption.value}']`
        );
        option.classList.remove("checked");
        selectedValueContainer.textContent = "";
        placeholderTextContainer.classList.remove("hidden");
        selectedValueContainer.classList.add("hidden");
        replaceIcon();
      }
    }
  });

  // Close dropdown when clicked outside
  document.addEventListener("click", (e) => {
    if (!selectContainer.contains(e.target)) {
      selectContainer.classList.remove("active");
    }
  });
})();

File: components/Select.go
package components

import (
	g "github.com/maragudk/gomponents"
	c "github.com/maragudk/gomponents/components"
	h "github.com/maragudk/gomponents/html"
)

type Option struct {
	Label g.Node
	Value string
}

func SelectOptions(o []Option, v []string, ID string) g.Node {
	return h.Ul(
		h.Class("dropdown"),
		h.Role("listbox"),
		h.ID(ID),
		g.Group(g.Map(o, func(o Option) g.Node {
			selected := false
			for _, val := range v {
				if val == o.Value {
					selected = true
					break
				}
			}

			return h.Div(
				c.Classes{
					"option":   true,
					"selected": selected,
				},
				h.Role("option"),
				h.DataAttr("value", o.Value),
				h.Span(h.Class("option-label"), o.Label),
				Icon(&IconProps{
					Identifier: "check",
				}),
			)
		})),
	)
}

type SelectProps struct {
	Options       []Option
	Value         []string
	Name          string
	ID            string
	LabelContents g.Node
	Multiple      bool
	Placeholder   string
	Classes       c.Classes
}

func Select(p *SelectProps) g.Node {
	if p.Classes == nil {
		p.Classes = c.Classes{}
	}

	p.Classes["select"] = true

	if p.Name == "" {
		panic("SelectProps.Name must be provided")
	}

	if p.ID == "" {
		p.ID = p.Name
	}

	if p.Placeholder == "" && !p.Multiple {
		p.Placeholder = "Select an option"
	} else if p.Placeholder == "" && p.Multiple {
		p.Placeholder = "Select options"
	}

	dropdownId := p.ID + "-dropdown"

	return h.Label(
		g.If(p.LabelContents != nil, p.LabelContents),
		h.Div(
			p.Classes,
			h.DataAttr("placeholder", p.Placeholder),
			h.DataAttr("name", p.Name),
			g.If(p.Multiple, h.DataAttr("multiple", "true")),
			h.ID(p.ID),
			h.Button(
				h.Role("combobox"),
				g.If(!p.Multiple, h.Aria("label", "select dropdown button")),
				h.Aria("haspopup", "listbox"),
				h.Aria("expanded", "false"),
				h.Aria("controls", dropdownId),

				// A span containing the placeholder/prompt
				h.Span(
					c.Classes{
						"placeholder": true,
						"hidden":      len(p.Value) > 0,
					},
					g.Text(p.Placeholder),
				),

				// A container for the currently selected values
				h.Span(
					h.Class("selected-values"),
					g.If(
						len(p.Value) > 0 && p.Multiple,
						g.Group(g.Map(p.Value, func(v string) g.Node {
							value := v
							var label g.Node
							for _, o := range p.Options {
								if o.Value == v {
									label = o.Label
									break
								}
							}
							return h.Span(
								h.Class("selected-value-chip"),
								h.DataAttr("value", value),
								label,
								h.Span(
									h.Class("remove"),
									g.Text("\u2715"),
								),
								h.Input(
									h.Type("checkbox"),
									h.Name(p.Name),
									h.Value(value),
									h.Checked(),
								),
							)
						})),
					),
					g.If(
						len(p.Value) > 0 && !p.Multiple,
						g.Group(g.Map(p.Value, func(v string) g.Node {
							value := v
							var label g.Node
							for _, o := range p.Options {
								if o.Value == v {
									label = o.Label
									break
								}
							}
							return h.Span(
								h.Class("selected-value"),
								h.DataAttr("value", value),
								label,
								h.Input(
									h.Type("radio"),
									h.Name(p.Name),
									h.Value(value),
									h.Disabled(),
									h.Checked(),
								),
							)
						})),
					),
				),

				// dropdown arrow
				h.Span(
					c.Classes{
						"dropdown-arrow": true,
						"hidden":         len(p.Value) > 0,
					},
					g.Text("\u25BE"), // small downwards pointing triangle
				),

				// clear icon
				h.Span(
					c.Classes{
						"clear-select": true,
						"hidden":       len(p.Value) == 0,
					},
					g.Text("\u2715"), // small close/clear icon
				),
			),
			SelectOptions(p.Options, p.Value, dropdownId),
		),
	)
}
